field/TinyLang/Field/UParser.hs

```diff
@@ -0,0 +1,258 @@
+analysis and grammar we will explicitly mark all lexical tokens.
+
+Please note that expressions are the designated start production for
+the parser.
+
+We are using the following convention
+
+* /tokens/ are enclosed in parentheses @""@.
+* /regular expressions/ are enclosed in square brackets @[]@.
+
+== Literals
+
+Literals are defined as follows
+
+@
+bool-literal ::= "T" | "F"
+vec-literal  ::= "{" (bool-literal ("," bool-literal)*)? "}"
+int-literal  ::= [0-9]+
+@
+
+== Keywords
+
+Here is a list of keywords used in the language.
+
+@
+keyword ::=
+  "and"
+  "assert"
+  "do"
+  "else"
+  "end"
+  "for"
+  "if"
+  "inv"
+  "let"
+  "neq0"
+  "not"
+  "or"
+  "then"
+  "to"
+  "unpack"
+  "xor"
+@
+
+== Identifiers
+
+To avoid a syntactic clash with /bool-literals/, identifiers for field
+variables start with a lower-case letter.  Identifiers for boolean
+variables are prefiex by @?@ and identifiers for vector variables are
+prefixed by @#@.  We will check if an identifier is not a keyword.
+
+@
+ident ::= ( "?" | "#" ) [a-z] ([a-z0-9_'])*
+@
+
+== Constants
+
+At the moment constants consist of literals only.
+
+@
+const ::=
+  bool-literal
+  int-literal
+  vec-literal
+@
+
+== Expressions
+
+As mentioned above @expr@ is the designated start production for the
+parser.
+
+@
+expr ::=
+  const
+  "(" expr ")"
+  ident
+  expr infix-op expr
+  prefix-op expr
+  expr "[" expr "]"
+  statement ";" expr
+  "if" expr "then" expr "else" expr
+
+infix-op ::=
+  "and"
+  "or"
+  "xor"
+  "=="
+  "<="
+  "<"
+  ">="
+  ">"
+  "+"
+  "-"
+  "*"
+  "/"
+
+prefix-op ::=
+  "not"
+  "neq0"
+  "neg"
+  "inv"
+  "unpack"
+@
+
+== Statement
+
+Statements are a bit odd at the moment, as they are neither
+expressions nor the usual statements.
+
+@
+statement ::=
+  "let" var "=" expr
+  "assert" expr
+  "for" var "=" int "to" int "do" statements "end"
+
+statements ::=
+  (statement (";" statement)*)?
+@
+
+== Operator Precedence
+  Precedence: "not" > "xor" > "and" > "or"  (but use parentheses anyway).
+  if-then-else has to be parenthesised unless it's at the very top.
+
+  Precedence for numeric operators is standard:  {neg,inv} > {*,/} > {+,- }.
+  Things like "neg inv 5" are illegal: use parentheses.
+
+  The code is based on the tutorial at
+  https://markkarpov.com/megaparsec/parsing-simple-imperative-language.html
+
+  See also https://markkarpov.com/megaparsec/megaparsec.html
+-}
+
+
+module TinyLang.Field.UParser where
+
+import           Control.Applicative.Combinators hiding (many, sepBy, some)
+import           Text.Megaparsec
+import           Text.Megaparsec.Char
+import qualified Text.Megaparsec.Char.Lexer      as L
+import           Data.Void
+import qualified Data.Set                        as S
+import           GHC.Unicode ( isLower
+                             , isDigit)
+
+type Parser = Parsec Void String
+
+type Identifier = String
+
+data Const
+  = CBool  Bool
+  | CInt   Integer
+  | CVec   [Bool]
+
+data Var = Var Identifier
```

Comment:
We don't need to define another `Var` here. Instead, `TinyLang.Var.Var` should be used.

Reply:
I was not sure about it.  `TinyLang.Var.Var` contains `Unique` field, which I believe is used to tell variables apart.

Since I am not checking for scopes at this stage, how should I generate values of `Unique`?

Reply:
In

```haskell
parseExpr
    :: forall f m. (MonadSupply m, TextField f)
    => String -> m (Either String (SomeUniExpr f))
parseExpr = fmap _scopedValue . parseScopedExpr
```

the `MonadSupply` constraint is responsible for generating fresh `Unique`s.

In `common/TinyLang/ParseUtils.hs` we have these definitions:

```
type Parser = ParsecT Void String (StateT Scope Supply)

-- | 'Scope' maps names onto 'Var's.
type Scope = Map String Var

data Scoped a = Scoped
    { _scopedScope :: Map String Var
    , _scopedValue :: a
    } deriving (Functor, Foldable, Traversable)

-- | Look up a variable name. If we've already seen it, return the corresponding Var;
-- otherwise, increase the Unique counter and use it to construct a new Var.
makeVar :: String -> Parser Var
makeVar name = do
    vars <- get
    case Map.lookup name vars of
        Just var -> pure var
        Nothing  -> do
            var <- freshVar name
            put $ Map.insert name var vars
            pure var
```

There are two options: either we keep these definitions or do not assign any `Unique`s to variables during parsing and instead do that in a separate pass, which forces us to parameterize the untyped AST by `var`, which gets instantiated to `String` during parsing and to `TinyLang.Var.Var` after the separate pass.

I'm slightly in favor of the latter option, because it also allows to instantiate `var` with a type of De Bruijn indices, which is handy for tests, compression, etc (we use that trick in Plutus). But the former option is fine too as it's fewer changes to make and transitioning from the former to the latter is a straightforward thing to do, so we can do that once we actually need it.

Reply:
Went with parameterising the AST by the type of the variable.